### PR TITLE
Clam 2296 coverity regressions

### PR DESCRIPTION
**Pull request by micahsnyder**
Fix a number of recent coverity issues introduced during 1.1 dev, along with a large number of compile time warnings:

- Fix many warnings

- Fix benign compiler warning in logging macro

- The strncpy intentionally is not copying the NULL terminator for the log message prefix. The NULL will be added by vsnprintf, after. Switching to memcpy eliminates the warning.

- Fix possibly uninitialized variable warning

- Coverity-401433: Fix file descriptor leak in VBA parser

- Coverity-401434: Switch to realloc that does not free after failure

- Because we free after the done-label.

- Coverity-404677: Add missing frees in sigtool

- Coverity-405726, 405725: Fix overlapping copy complaint

- Fix issue introduced during 1.1 dev.

- Fix coverity-405726 coverity-405725.

- Coverity-405733, 405732: Add missing variable initializers

- Coverity-405734: Add missing variable initializer

- Coverity-405735: Add missing 'goto done;' on error

### CODE REVIEW

File: libclamav/scanners.c [hunk 1]
```diff
@@ -2045,6 +2045,10 @@ static cl_error_t cli_ole2_tempdir_scan_vba(const char *dir, cli_ctx *ctx, struc
         free(fullname);
     }
 
+    if (fd >= 0) {
```
Comment (ragusaa) [inline libclamav/scanners.c:2048]: shouldn't this be > 2, since 0 is stdin, 1 is stdout, and 2 is stderr?

**Review by ragusaa — CHANGES_REQUESTED** [review]
Code looks good, just made a few suggestions.

File: libclamav/dsig.c [hunk 2]
```diff
@@ -417,6 +419,7 @@ int cli_versig2(const unsigned char *sha256, const char *dsig_str, const char *n
     return memcmp(digest1, digest2, HASH_LEN) ? CL_EVERIFY : CL_SUCCESS;
 
 done:
+    free(decoded);
```
[the remark this answers is not in the document]
Comment (micahsnyder) [reply libclamav/dsig.c:422]: I used to be superstitious about this, but from what I can tell it's safe everywhere if the pointer is NULL.
E.g. on Windows: https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/free?view=msvc-170

File: libclamav/explode.c [hunk 3]
```diff
@@ -198,8 +198,8 @@ int explode_init(struct xplstate *X, uint16_t flags)
         }                                                                 \
     }
 
-#define GETCODES(CASE, WHICH, HOWMANY)                                        \
-    case CASE: {                                                              \
+#define GETCODES(WHICH, HOWMANY)                                              \
```
[the remark this answers is not in the document]
Comment (micahsnyder) [reply libclamav/explode.c:201]: This is the one I messaged you about, and then was like "nevermind I thought of something". 
It was the only way I could figure out how to get GCC to recognize the "fall-through" comment. But I think I also like it.

File: libclamav/scanners.c (hunk 1 above)
Comment (micahsnyder) [reply libclamav/scanners.c:2048]: It's possible to close stdin, stdout, and stderr in which case the 0-2 are valid file descriptors for files.  Not that we do that, as far as I know.  But perhaps some application using libclamav might?

File: clamav-milter/clamav-milter.c [hunk 4]
```diff
@@ -175,10 +175,10 @@ int main(int argc, char **argv)
 
     pt = optget(opts, "AddHeader")->strarg;
     if (strcasecmp(pt, "No")) {
-        char myname[255];
```
[the remark this answers is not in the document]
Comment (micahsnyder) [reply clamav-milter/clamav-milter.c:178]: I see this warning with this change:
```

/home/micah/workspace/clamav-micah/clamav-milter/clamav-milter.c: In function ‘main’:
/home/micah/workspace/clamav-micah/clamav-milter/clamav-milter.c:185:73: warning: ‘%s’ directive output may be truncated writing up to 254 bytes into a region of size 237 [-Wformat-truncation=]
  185 |             snprintf(xvirushdr, sizeof(xvirushdr), "clamav-milter %s at %s",
      |                                                                         ^~
  186 |                      get_version(), myname);
      |                                     ~~~~~~
/home/micah/workspace/clamav-micah/clamav-milter/clamav-milter.c:185:13: note: ‘snprintf’ output 19 or more bytes (assuming 273) into a destination of size 255
  185 |             snprintf(xvirushdr, sizeof(xvirushdr), "clamav-milter %s at %s",
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  186 |                      get_version(), myname);
      |                      ~~~~~~~~~~~~~~~~~~~~~~
      
```
So to resolve that, perhaps I need to increase the size of the `xvirushdr` buffer. I'll add this:
```
--- a/clamav-milter/clamfi.c
+++ b/clamav-milter/clamfi.c
@@ -61,7 +61,7 @@ static sfsistat (*InfectedAction)(SMFICTX *ctx);
 static char *rejectfmt = NULL;

 int addxvirus = 0; /* 0 - don't add | 1 - replace | 2 - add */
-char xvirushdr[255];
+char xvirushdr[300];
 char *viraction = NULL;
 int multircpt   = 1;

diff --git a/clamav-milter/clamfi.h b/clamav-milter/clamfi.h
index 61e18fbef..08134277e 100644
--- a/clamav-milter/clamfi.h
+++ b/clamav-milter/clamfi.h
@@ -27,7 +27,7 @@

 extern uint64_t maxfilesize;
 extern int addxvirus;
-extern char xvirushdr[255];
+extern char xvirushdr[300];
 extern int multircpt;

 sfsistat clamfi_body(SMFICTX *ctx, unsigned char *bodyp, size_t len);

```

**Pull request merged**